experimental/devices/mfrc522/mfrc522.go

```diff
@@ -174,68 +177,192 @@ func (r *Dev) Halt() error {
-		if err != nil {
-			return err
+	defer func() {
+		r.accessFinished()
+		if err == nil {
+			err = r.stopCrypto()
 		}
-		if current&0x03 != 0 {
-			return wrapf("can not set the bitmask for antenna")
+	}()
+	uuid, err := r.selectCard()
+	if err != nil {
+		return
+	}
+
+	state, err := r.auth(auth, calcBlockAddress(sector, 3), key, uuid)
+	if err != nil {
+		return
+	}
+	if state != AuthOk {
+		return nil, wrapf("can not authenticate")
+	}
+
+	return r.read(calcBlockAddress(sector, 3))
+}
+
+// WriteCard writes the data into the card block.
+//
+// 	auth - the authentiction mode.
+// 	sector - the sector on the card to write to.
+// 	block - the block within the sector to write into.
+// 	data - 16 bytes if data to write
+// 	key - the key used to authenticate the card - depends on the used auth method.
+func (r *Dev) WriteCard(auth byte, sector int, block int, data [16]byte, key [6]byte) (err error) {
+	if err = r.accessStarted(); err != nil {
+		return
+	}
+
+	defer func() {
+		r.accessFinished()
+		if err == nil {
+			err = r.stopCrypto()
 		}
-		return r.setBitmask(commands.TxControlReg, 0x03)
+	}()
+	uuid, err := r.selectCard()
+	if err != nil {
+		return
 	}
-	return r.clearBitmask(commands.TxControlReg, 0x03)
+	state, err := r.auth(auth, calcBlockAddress(sector, 3), key, uuid)
+	if err != nil {
+		return
+	}
+	if state != AuthOk {
+		err = wrapf("authentication failed")
+		return
+	}
+
+	return r.write(calcBlockAddress(sector, block%3), data[:])
 }
 
-// SetAntennaGain configures antenna signal strength.
+// WriteSectorTrail writes the sector trail with sector access bits.
 //
-//	gain - signal strength from 0 to 7.
-func (r *Dev) SetAntennaGain(gain int) {
-	r.aMu.Lock()
-	defer r.aMu.Unlock()
+// 	auth - authentication mode.
+// 	sector - sector to set authentication.
+// 	keyA - the key used for AuthA authentication scheme.
+// 	keyB - the key used for AuthB authentication scheme.
+// 	access - the block access structure.
+// 	key - the current key used to authenticate the provided sector.
+func (r *Dev) WriteSectorTrail(auth byte, sector int, keyA [6]byte, keyB [6]byte, access *BlocksAccess, key [6]byte) (err error) {
+	if err = r.accessStarted(); err != nil {
+		return
+	}
+
+	defer func() {
+		r.accessFinished()
+		if err == nil {
+			err = r.stopCrypto()
+		}
+	}()
+	uuid, err := r.selectCard()
+	if err != nil {
+		return
+	}
+	state, err := r.auth(auth, calcBlockAddress(sector, 3), key, uuid)
+	if err != nil {
+		return
+	}
+	if state != AuthOk {
+		err = wrapf("failed to authenticate")
+		return
+	}
+
+	var data [16]byte
+	copy(data[:], keyA[:])
+	accessData := CalculateBlockAccess(access)
+	copy(data[6:], accessData[:4])
+	copy(data[10:], keyB[:])
+	return r.write(calcBlockAddress(sector&0xFF, 3), data[:])
+}
 
-	if 0 <= gain && gain <= 7 {
-		r.antennaGain = gain
+//		MFRC522 SPI Dev private/helper functions
+
+// Checks whether device is already in use.
+func (r *Dev) accessStarted() error {
```

Comment:
Is it possible to take a function as an argument to and wrap the execution with lock/unlock?

this way it will be more clean to see the boundaries of the invocations, and less error-prone to forget to unlock the mutex.

Something like 
```
func withDeviceAccess(f func() error) error {
  r.xMu.Lock()
  defer r.Xmu.Unlock()
  return f()
}
```

Reply:
It's more about the style, I would say. Most of the commands that anyone would need to implement are atomic, e.g - a single thread gets access to the device and performs a sequence of the commands on it. So it is better to have a function that wraps that sequence in "transaction" with locking the library, and pass a function that does something useful.

I could be wrong, but the current implementation you have to call
- accessStarted
- do something
- accessStopped

Here we have 3 blocks and it is fairly easy to panic in the code and have accessStopped lock the device forever.

So with the "pass the function and wrap the context" we are guaranteed to have the library lock managed in the single place, and we don't need an auxiliary `bool` flag to represent the device state.

Hope that it makes sense.

Reply:
Yes. That said I'm also fine with:

```
r.lockDevice()
defer r.unlockDevice()
```

which essentially boils down to the same thing (unless mistaken?) and I think readers can relate to this as a lock. Basically rename the function and remove the error.

Reply:
The reason I'm not using explicit lock is that Read could be a blocking operation, for example, if timeout set for indefinite wait second thread will be locked until a card is detected in a first thread.

Reply:
Ok, let's leave it as that, please still fix the issues in example_test.go and it's fine to go.